hwtracer/src/perf/collect.rs

```diff
@@ -137,16 +137,17 @@ impl PerfThreadTracer {
 
 /// A wrapper around a manually malloc/free'd buffer for holding an Intel PT trace. We've split
 /// this out from PerfTrace so that we can mark just this raw pointer as `unsafe Send`.
+///
+/// FIXME: Marking a pointer `Send` is necessary but not sufficient: we also need some sort of
+/// memory barrier to guarantee that all writes from thread A are seen by thread B. That said, x86
+/// (remember that this code can only run on x86) has a strong memory model makes it relatively
```

Comment:
In theory this part of the code isn't platform specific. We don't have to worry about it now, because we don't support other arches, but I think we'd probably still [use perf to collect (e.g.) coresight traces](https://docs.kernel.org/trace/coresight/coresight-perf.html).

It's been a while since I've thought about concurrent memory model stuff. Is the problematic read/write the pointer value itself, or what it points to?

Reply:
What it points to.

FWIW, `hwtracer` probably needs to be rewritten from scratch before we can port it. We've learnt so much since we originally wrote this, and the internal structure has now been pushed and pulled so much that I find working with it very difficult. We can probably let it retire with grace at some point and make a simpler, safer, alternative. But not yet :)

Reply:
Agreed.

Can we just change:

> remember that this code can only run on x86

To something like:

> currently the system only supports x86, so...

And then hopefully someone will come along re-write hwtracer in rust and we can all retire :)

Reply:
> What it points to.

Maybe it isn't possible, but could we not use a memory clobber somewhere?

Reply:
It's probably possible, but I wasn't even quite sure what the right place to do this is. I think working that out might be in danger of throwing good money after bad.

Reply:
ok

Reply:
(so just the comment re-word then?)

Reply:
Fixed in bf6f6585. Also added a `cfg` guard for x86 to enforce the comment!

Reply:
bf6f6585 has now shown up